pgjdbc/src/main/java/org/postgresql/core/v3/ConnectionFactoryImpl.java

```diff
@@ -864,9 +866,23 @@ private void runInitialQueries(QueryExecutor queryExecutor, Properties info)
   }
 
   private boolean isPrimary(QueryExecutor queryExecutor) throws SQLException, IOException {
-    Tuple results = SetupQueryRunner.run(queryExecutor, "show transaction_read_only", true);
+    String inHotStandby = queryExecutor.getParameterStatus(IN_HOT_STANDBY);
+    if ("on".equalsIgnoreCase(inHotStandby)) {
+      return false;
+    }
+    // Host may not be in recovery but still have 'transaction_read_only=on' set,
+    // hence explicit check
+    Tuple results = SetupQueryRunner.run(queryExecutor, showIsPrimaryGuc(queryExecutor), true);
     Tuple nonNullResults = castNonNull(results);
-    String value = queryExecutor.getEncoding().decode(castNonNull(nonNullResults.get(0)));
-    return value.equalsIgnoreCase("off");
+    String queriedTransactionReadonly = queryExecutor.getEncoding().decode(castNonNull(nonNullResults.get(0)));
+    return queriedTransactionReadonly.equalsIgnoreCase("off");
+  }
+
+  private String showIsPrimaryGuc(QueryExecutor queryExecutor) {
```

Comment:
Docs says `in_hot_standby` is more accurate way to determine server status (primary/replica). On the other hand pgjdbc users may have gotten used to the idea that "primary" == "can't issue write transactions" ie 
> if `targetServerType=primary/master` then connection established to the server where `transaction_read_only` is `off/false`

Reply:
Since the server has created a dedicated GUC for `inHotStandby` clearly that comment was "optimistic". But it was all we had. I don't think there is any reason to make the comment terse. Please use as much verbiage as possible to explain it. I think noting that it is possible for a primary to have `transaction_read_only=off` if the user manually set it

Reply:
Done.